primitives/npos-elections/solution-type/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+/// type, `u8` target type and `Perbill` accuracy with maximum of 4 edges per voter.
+///
+/// ```
+/// # use sp_npos_elections_solution_type::generate_solution_type;
+/// # use sp_arithmetic::per_things::Perbill;
+/// generate_solution_type!(pub struct TestSolution::<
+///     VoterIndex = u16,
+///     TargetIndex = u8,
+///     Accuracy = Perbill,
+/// >(4));
+/// ```
+///
+/// The output of this macro will roughly look like:
+///
+/// ```ignore
+/// struct TestSolution {
+/// 	voters1: vec![(u16 /* voter */, u8 /* target */)]
+/// 	voters2: vec![
+/// 		(u16 /* voter */, [u8 /* first target*/, Perbill /* proportion for first target */], u8 /* last target */)
+/// 	]
+/// 	voters3: vec![
+/// 		(u16 /* voter */,  [
+/// 			(u8 /* first target*/, Perbill /* proportion for first target */ ),
+/// 			(u8 /* second target */, Perbill /* proportion for second target*/)
+/// 		], u8 /* last target */)
+///		],
+/// 	voters4: ...,
+/// }
+///
+/// impl NposSolution for TestSolution {};
+/// impl Solution for TestSolution {};
+/// ```
+///
+/// The given struct provides function to convert from/to `Assignment` as part of
+/// [`sp_npos_elections::Solution`] trait:
+///
+/// - `fn from_assignment<..>(..)`
+/// - `fn into_assignment<..>(..)`
+///
+/// ## Compact Encoding
+///
+/// The generated struct is by default deriving both `Encode` and `Decode`. This is okay but could
+/// lead to many `0`s in the solution. If prefixed with `#[compact]`, then a custom compact encoding
+/// for numbers will be used, similar to how `parity-scale-codec`'s `Compact` works.
+///
+/// ```
+/// # use sp_npos_elections_solution_type::generate_solution_type;
+/// # use sp_npos_elections::Solution;
+/// # use sp_arithmetic::per_things::Perbill;
+/// generate_solution_type!(
+///     #[compact]
+///     pub struct TestSolutionCompact::<VoterIndex = u16, TargetIndex = u8, Accuracy = Perbill>(8)
+/// );
+/// ```
+#[proc_macro]
+pub fn generate_solution_type(item: TokenStream) -> TokenStream {
+	let solution_def = syn::parse_macro_input!(item as SolutionDef);
+
+	let imports = imports().unwrap_or_else(|e| e.to_compile_error());
+
+	let def = single_page::generate(solution_def).unwrap_or_else(|e| e.to_compile_error());
+
+	quote!(
+		#imports
+		#def
+	)
+	.into()
+}
+
+struct SolutionDef {
+	vis: syn::Visibility,
+	ident: syn::Ident,
+	voter_type: syn::Type,
+	target_type: syn::Type,
+	weight_type: syn::Type,
+	count: usize,
+	compact_encoding: bool,
+}
+
+fn check_attributes(input: ParseStream) -> syn::Result<bool> {
+	let attrs = input.call(syn::Attribute::parse_outer).unwrap_or_default();
+	if attrs.len() > 1 {
+		return Err(syn_err("compact solution can accept only #[compact]"))
+	}
+
+	Ok(attrs.iter().any(|attr| {
+		if attr.path.segments.len() == 1 {
+			let segment = attr.path.segments.first().expect("Vec with len 1 can be popped.");
+			if segment.ident == Ident::new("compact", Span::call_site()) {
+				return true
+			}
+		}
```

Comment:
those both if should have an else which returns an error.
The thing is that either there is no attribute or the attribute must be `#[compact]`. Currently it return false if the attribute is misswritten.
Like `#[compacts]` does not return error but instead ignore the attribute.

Anyway this is not introduced by this PR as far as I know.

Reply:
made an issue for it.